llvmPackages_12: build from monorepo source

sternenseemann opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

As mentioned in https://github.com/NixOS/nixpkgs/issues/305146, keeping LLVM 12 is a source of pain because it
is the only version to be built from individual release tarball instead
of the LLVM monorepo. This change makes LLVM 12 start from the monorepo
as well, simplifying all common LLVM expressions in the process.
No more LLVM 12 specific `default.nix`!

Additionally, some additional, simple cleanups are included that just rebuild LLVM 12.
Consequently, this PR is best reviewed commit by commit.
The common `default.nix`'s patch logic can further be simplified by a separate change
that rebuilds all LLVM versions.

This change is not breaking and could be included in 24.11 (or backported if we miss
the window) which seems wise. I've tested LLVM 12 on x86_64-linux and aarch64-linux.
I've also tested GHC 8.10.7 (which uses LLVM 12) on aarch64-linux.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

sternenseemann committed:
llvmPackages_12: build from monorepo source

As mentioned in #305146, keeping LLVM 12 is a source of pain because it
is the only version to be built from individual release tarball instead
of the LLVM monorepo. This commit makes LLVM 12 start from the monorepo
as well, simplifying all common LLVM expressions in the process.

With #347887, some quirks in the expressions for LLVM <14 were ironed
out, so building LLVM through from the monorepo is quite simple now.

- Most expressions only required minor changes, mostly removing the
  special casing for `sourceRoot`.

- The patch lists from llvm/12/default.nix were ported to
  common/default.nix. This only required a few extra conditionals which
  could be reduced via a rebuild also involving other LLVM versions.
  Outstanding tasks of little urgency have been noted in TODO comments.
  I have verified that the patch lists stay the same for all packages
  except LLVM where merely the order changes. An extra set of eyes
  is appreciated, of course.

- clang: The expression was reworked to use the same symlink location
  for clang-tools-extra for all versions including LLVM 12. This
  required adjusting the ad hoc patching of the clangd cmake files
  slightly.

- libunwind: We no longer need to make the libcxx sources available
  manually. We can rely on the monorepo source instead.

- lld: We no longer need to make the libunwind sources available manually.

- llvm: We no longer need to make the polly sources available manually

- On Darwin, we need to bypass CMake's C++ compiler for libcxx and
  libunwind now. It isn't a 100% clear why, probably because we've
  started to use Darwin's bootstrapStdenv for libcxx in the common
  expression compared to LLVM 12 on master [1].
  The reordering of flags for wasm causes a rebuild for some packages
  like firefox, but this should be tolerable on staging.

[1]: https://github.com/NixOS/nixpkgs/blob/665ebfb253caba7b85c2affefe2a92b305def4e6/pkgs/development/compilers/llvm/12/default.nix#L392-L430

sternenseemann committed:
llvmPackages_12.compiler-rt: move codesign patch into versioned dir

All LLVM versions < 12 have been removed, so this patch can live in the
versioned directory, simplifying the patch list in the process.

sternenseemann committed:
llvmPackages_12.clang: use nostdlibinc patch instead of sed command

Seems like the patch we've written for LLVM 13 and above also works for
LLVM 12 which seems a little more robust.

RossComputerGuy commented:
`pkgs/development/compilers/llvm/12/default.nix` is gone 🎉 

`pkgs/development/compilers/llvm/common/llvm/default.nix` has conflicts.

sternenseemann commented:
I'm going to re-test everything, now that everything is rebased on staging, just to be sure.

RossComputerGuy commented:
lldb in LLVM 12 failed to resolve the `cpu_subtype_arm64e_replacement` patch.

sternenseemann committed:
llvmPackages_{12,13}.lldb: don't try to find nonexistent patch

We just need to make sure the attribute evaluates so we can cleanly
access `meta.broken` for those versions.

sternenseemann commented:
`x86_64-linux` builds fine, but `aarch64-darwin` doesn't. It appears that it is already broken in some way before this change. Is anyone aware of any details?

sternenseemann commented:
@reckenrode Any ideas about the followng error (staging, aarch64-darwin, llvmPackages_12.libcxx)?

```
-- Check for working CXX compiler: /nix/store/7qfjlhi98568y1ff5xy93607dgn3lfw8-clang-wrapper-12.0.1/bin/clang++
-- Check for working CXX compiler: /nix/store/7qfjlhi98568y1ff5xy93607dgn3lfw8-clang-wrapper-12.0.1/bin/clang++ - broken
CMake Error at /nix/store/xl246ny58q2xm0w7srsjkkd1j3f7hakg-cmake-3.30.4/share/cmake-3.30/Modules/CMakeTestCXXCompiler.cmake:73 (message):
  The C++ compiler

    "/nix/store/7qfjlhi98568y1ff5xy93607dgn3lfw8-clang-wrapper-12.0.1/bin/clang++"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: '/tmp/nix-build-libcxx-12.0.1.drv-0/libcxx-src-12.0.1/runtimes/build/CMakeFiles/CMakeScratch/TryCompile-Ul4rVU'
    
    Run Build Command(s): /nix/store/6qaz2a5vrp8bv8wyjcyfkmssxdcbdc2d-ninja-1.12.1/bin/ninja -v cmTC_70287
    [1/2] /nix/store/7qfjlhi98568y1ff5xy93607dgn3lfw8-clang-wrapper-12.0.1/bin/clang++   -isysroot /nix/store/9qf26qdqy34chms1vx44w2cm7mjg46nh-apple-sdk-11.3/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -MD -MT CMakeFiles/cmTC_70287.dir/testCXXCompiler.cxx.o -MF CMakeFiles/cmTC_70287.dir/testCXXCompiler.cxx.o.d -o CMakeFiles/cmTC_70287.dir/testCXXCompiler.cxx.o -c /tmp/nix-build-libcxx-12.0.1.drv-0/libcxx-src-12.0.1/runtimes/build/CMakeFiles/CMakeScratch/TryCompile-Ul4rVU/testCXXCompiler.cxx
    [2/2] : && /nix/store/7qfjlhi98568y1ff5xy93607dgn3lfw8-clang-wrapper-12.0.1/bin/clang++ -isysroot /nix/store/9qf26qdqy34chms1vx44w2cm7mjg46nh-apple-sdk-11.3/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -Wl,-search_paths_first -Wl,-headerpad_max_install_names  CMakeFiles/cmTC_70287.dir/testCXXCompiler.cxx.o -o cmTC_70287   && :
    FAILED: cmTC_70287 
    : && /nix/store/7qfjlhi98568y1ff5xy93607dgn3lfw8-clang-wrapper-12.0.1/bin/clang++ -isysroot /nix/store/9qf26qdqy34chms1vx44w2cm7mjg46nh-apple-sdk-11.3/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -Wl,-search_paths_first -Wl,-headerpad_max_install_names  CMakeFiles/cmTC_70287.dir/testCXXCompiler.cxx.o -o cmTC_70287   && :
    ld: library not found for -lc++
    clang-12: error: linker command failed with exit code 1 (use -v to see invocation)
    ninja: build stopped: subcommand failed.
    
    

  

  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  CMakeLists.txt:3 (project)
```

Interestingly, before my PR, compilation fails later, in `compiler-rt-libc`. That error is also a mystery to me.

reckenrode commented:
> Interestingly, before my PR, compilation fails later, in compiler-rt-libc. That error is also a mystery to me.

Darwin needed to use `buildLlvmPackages.clangWithLibcAndBasicRtAndLibcxx`, which it wasn’t before in the separate derivation. It does in the common one.

> Any ideas about the followng error (staging, aarch64-darwin, llvmPackages_12.libcxx)?

Its test for a working C++ compiler appears to be trying to link libc++ even though libcxx is building with `buildLlvmPackages.clangWithLibcAndBasicRt`, which does not have libc++. Before the derivations were unified, libcxx on Darwin built with a plan stdenv. Does it build if you define `"-DCMAKE_CXX_COMPILER_WORKS=ON"` on Darwin for libcxx-12?

sternenseemann commented:
> Does it build if you deaddine `"-DCMAKE_CXX_COMPILER_WORKS=ON"` to `cmakeFlags` on Darwin for libcxx-12?

Yes, that work and makes sense. What confuses me is that the same appears to be needed for libunwind as well which uses `clangWithLibcAndBasicRt` both before and after the change. I suppose this is from `libcxx` CMake files being pulled in…

sternenseemann commented:
`libcxx` wasn't built with libc++ available from what I can tell:

https://github.com/NixOS/nixpkgs/blob/665ebfb253caba7b85c2affefe2a92b305def4e6/pkgs/development/compilers/llvm/12/default.nix#L421

The difference seems to be that we use the bootstrap stdenv now:

https://github.com/NixOS/nixpkgs/blob/0d7fbcdaa1d959bac0ade9eff3c256a5caac050a/pkgs/development/compilers/llvm/common/default.nix#L1129-L1133

But for libunwind I can't find such a difference.

sternenseemann commented:
Current state of the PR builds successfully on x86_64-linux, aarch64-darwin and aarch64-linux. It bugs me a bit that I don't quite know why bypassing the check is necessary all of a sudden, but sometimes you have to accept these things I guess…

reckenrode commented:
> But for libunwind I can't find such a difference.

Darwin uses the system unwinder by default, so it can use the standard `stdenv` to build libunwind.

The `bootstrapStdenv` provides a Darwin SDK without any propagated inputs. It’s used to avoid infinite recursions (typically when cross-compiling). In LLVM’s case, it needs to use the `bootstrapStdenv` to build libc++ and compiler-rt because those packages are transitive dependencies of `xcrun`, which is propagated by the SDK (the new Darwin `stdenv` includes `xcrun` by default).

emilazy commented:
Looks like this rebuilds all of Haskell on AArch64, which is weird as I don’t think our default GHC version should be using LLVM 12?

Since we’re landing `haskell-updates` in `staging-next` anyway and this is the last 24.11-pre cycle, I think it would be fine to rebase on `staging-next` and land it there.

sternenseemann commented:
GHC 9.6 is bootstrapped using GHC 9.2 which uses LLVM 12.

emilazy commented:
OK. I believe we haven’t yet done an eval with the `haskell-updates` merge on `staging-next` anyway, so as long as this is rebased soon I think it should be fine to land directly in `staging-next` for 24.11. (Though I’m double checking in the Staging room to make sure I’m reading Hydra right.)

sternenseemann commented:
Rebased, but currently untested. I'd at least want to look at ofborg's rebuild overview before merging.

Ericson2314 commented:
Glad this is happening! :)

emilazy commented:
Builds on the community builders, only 220 rebuilds outside of `aarch64-linux`, and those look as expected for being downstream of Haskell packages that we just merged a huge rebuild of anyway. Going to merge this into `-next`.

emilazy commented:
Realized that actually the `haskell-updates` had already hit `staging-next` and that the revert/unrevert dance was just because it hit the trunk job a few days later. So I’m going to merge this into `staging` for 25.05 after all and we can backport it after branch‐off.

sternenseemann commented:
I had a failure in `llvmPackages_12.compiler-rt` on aarch64-darwin about a symlink failing to be created because it already existed, but I can't reproduce it. I suppose there's a race condition in the build system.

paparodeo commented:
> I had a failure in `llvmPackages_12.compiler-rt` on aarch64-darwin about a symlink failing to be created because it already existed, but I can't reproduce it. I suppose there's a race condition in the build system.

https://github.com/llvm/llvm-project/commit/b31080c596246bc26d2493cfd5e07f053cf9541c looks like it fixes the race

pwaller commented on pkgs/development/compilers/llvm/common/clang/default.nix:
```diff
@@ -79,17 +77,12 @@ let
+      (cd tools && ln -s ../../clang-tools-extra extra)
+    '' + lib.optionalString (lib.versionOlder release_version "13") ''
+      substituteInPlace tools/extra/clangd/quality/CompletionModel.cmake \
+        --replace ' ''${CMAKE_SOURCE_DIR}/../clang-tools-extra' ' ''${CMAKE_SOURCE_DIR}/tools/extra'
```
@sternenseemann 

https://github.com/NixOS/nixpkgs/pull/356120 (--replace-fail) broke this one on clang 12 (staging 03ed65c58212fded6339d4e59167d2d8765ae0d5).

The path tools/extra doesn't exist; the only reference to it I could find is in the deleted diff hunk which appears after this one. I take it you somehow separated out clang-tools-extra from this derivation and therefore it is vestigial and can be removed. Correct?

pwaller replied:
Oh no, wait, I see `(cd tools && ln -s ../../clang-tools-extra extra)` reatined in this diff which I missed. Hmm.

pwaller replied:
Found it, it's vestigial because the patch which removed CMAKE_SOURCE_DIR was removed in 12.0.1:
https://github.com/llvm/llvm-project/commit/3be5dbbc32dd348dddcec9e3baec96b7dcef8a35

sternenseemann replied:
Can't comment much more, I've just tried to preserve all the parts of the phases in the original LLVM 12 expressions. They may have been unnecessarily copied from earlier versions, yes.

pwaller replied:
Fix for llvm 12: https://github.com/NixOS/nixpkgs/pull/367056
